File: DropDownTableViewObjC/DropDownTableViewObjC/TestViewController.m
//
//  TestViewController.m
//  DropDownTableViewObjC
//
//  Created by NSSimpleApps on 15.01.17.
//  Copyright © 2017 NSSimpleApps. All rights reserved.
//

#import "TestViewController.h"

@interface NSDate (CustomFormat)

- (NSString *)customDateFormat;

@end

@implementation NSDate (CustomFormat)

- (NSString *)customDateFormat {
    
    NSDateFormatter *dateFormatter = [NSDateFormatter new];
    dateFormatter.dateFormat = @"d MMM YYYY";
    
    return [dateFormatter stringFromDate:self];
}

@end

@interface TestViewController () <UIPickerViewDataSource, UIPickerViewDelegate>

@property (assign) NSInteger value;
@property (strong) NSDate *date;

@end

@implementation TestViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
        
    self.title = @"Test";
        
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
        
    self.value = [userDefaults integerForKey:@"value"];
    
    self.date = [NSDate date];
    
    NSDate *date = [userDefaults objectForKey:@"date"];
    
    if (date && [date isKindOfClass:[NSDate class]]) {
        
        self.date = date;
        
    } else {
        
        self.date = [NSDate date];
    }
}

- (void)dealloc {
    
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setInteger:self.value forKey:@"value"];
    [userDefaults setObject:self.date forKey:@"date"];
}

- (NSInteger)numberOfRowsInTableView:(UITableView *)tableView {
    
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfSubrowsInRow:(NSInteger)row {
    
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    if (row == 0) {
        
        cell.textLabel.text = @"Value";
        cell.detailTextLabel.text = @(self.value).stringValue;
        
    } else {
        
        cell.textLabel.text = @"Date";
        cell.detailTextLabel.text = self.date.customDateFormat;
    }
    
    if (self.nsk_selectedRow) {
        
        NSInteger selectedRow = self.nsk_selectedRow.integerValue;
        
        if (row == selectedRow) {
            
            cell.accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
            
        } else {
            
            cell.accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
        }
    }
    
    return cell;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForSubrow:(NSInteger)subrow inRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell;
    
    if (row == 0) {
        
        cell = [tableView dequeueReusableCellWithIdentifier:@"PickerCell" forIndexPath:indexPath];
        
        // this is dirty hack. Don't do it anymore
        UIPickerView *pickerView = cell.contentView.subviews.firstObject;
        [pickerView selectRow:self.value inComponent:0 animated:NO];
        
    } else {
        
        cell = [tableView dequeueReusableCellWithIdentifier:@"DateCell" forIndexPath:indexPath];
        
        // this is dirty hack. Don't do it anymore
        UIDatePicker *datePicker = cell.contentView.subviews.firstObject;
        datePicker.date = self.date;
    }
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    return 216;
}

- (void)tableView:(UITableView *)tableView didSelectRow:(NSInteger)row {
    
    if (self.nsk_selectedRow) {
        
        NSInteger selectedRow = self.nsk_selectedRow.integerValue;
        
        if (row == selectedRow) {
            
            [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
            [tableView deselectRow:row animated:YES];
            
        } else {
            
            [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
            [tableView cellForRowAtRow:selectedRow].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
        }
        
    } else {
        
        [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
    }
    
    [super tableView:tableView didSelectRow:row];
}

- (IBAction)dateChanged:(UIDatePicker *)sender {
        
    self.date = sender.date;
        
    UITableViewCell *cell = [self.tableView cellForRowAtRow:1];
    
    if (cell) {
        
        cell.detailTextLabel.text = self.date.customDateFormat;
    }
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    
    return 10;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    
    return @(row).stringValue;
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    
    self.value = row;
    
    UITableViewCell *cell = [self.tableView cellForRowAtRow:0];
    
    if (cell) {
        
        cell.detailTextLabel.text = @(self.value).stringValue;
    }
}


@end

File: DropDownTableViewSwift/DropDownTableViewSwift/TestViewController.swift
//
//  TestViewController.swift
//  DropDownTableView
//
//  Created by NSSimpleApps on 26.12.15.
//  Copyright © 2015 NSSimpleApps. All rights reserved.
//

import UIKit

private extension Date {
    
    var customDateFormat: String {
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "d MMM YYYY"
        
        return dateFormatter.string(from: self)
    }
}

class TestViewController: DropDownTableViewController {
    
    fileprivate var value = 0
    fileprivate var date = Date()
    
    override func viewDidLoad() {
        
        super.viewDidLoad()

        self.title = "Test"
        
        let userDefaults = UserDefaults.standard
        
        self.value = userDefaults.integer(forKey: "value")
        
        if let date = userDefaults.object(forKey: "date") as? Date {
            
            self.date = date
        }
    }

    deinit {
        
        let userDefaults = UserDefaults.standard
        userDefaults.set(self.value, forKey: "value")
        userDefaults.set(self.date, forKey: "date")
    }
    
    override func numberOfRows(in tableView: UITableView) -> Int {
        
        return 2
    }
    
    override func tableView(_ tableView: UITableView, numberOfSubrowsInRow row: Int) -> Int {
        
        return 1
    }
    
    override func tableView(_ tableView: UITableView, cellForRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        
        if row == 0 {
            
            cell.textLabel?.text = "Value"
            cell.detailTextLabel?.text = String(self.value)
            
        } else {
            
            cell.textLabel?.text = "Date"
            cell.detailTextLabel?.text = self.date.customDateFormat
        }
        
        if row == self.nsk_selectedRow {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            
        } else {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, cellForSubrow subrow: Int, inRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let cell: UITableViewCell
        
        if row == 0 {
            
            cell = tableView.dequeueReusableCell(withIdentifier: "PickerCell", for: indexPath)
            
            // this is dirty hack. Don't do it anymore
            if let pickerView = cell.contentView.subviews.first as? UIPickerView {
                
                pickerView.selectRow(self.value, inComponent: 0, animated: false)
            }
            
        } else {
            
            cell = tableView.dequeueReusableCell(withIdentifier: "DateCell", for: indexPath)
            
            // this is dirty hack. Don't do it anymore
            if let datePicker = cell.contentView.subviews.first as? UIDatePicker {
                
                datePicker.date = self.date
            }
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRow row: Int) {
        
        switch (self.nsk_selectedRow, row) {
            
        case (let sr?, _) where row == sr:
            tableView.cellForRow(at: row)?.accessoryView =  UIImageView(image: UIImage(named: "down_arrow"))
            tableView.deselect(row: row, animated: true)
            break
            
        case (let sr?, _) where row != sr:
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            tableView.cellForRow(at: sr)?.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
            break
            
        case (nil, _):
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            break
            
        default:
            break
        }
        
        super.tableView(tableView, didSelectRow: row)
    }
    
    override func tableView(_ tableView: UITableView, heightForSubrow subrow: Int, inRow row: Int) -> CGFloat {
        
        return 216
    }
    
    @IBAction func dateChanged(_ sender: UIDatePicker) {
        
        self.date = sender.date
        
        if let cell = self.tableView.cellForRow(at: 1) {
            
            cell.detailTextLabel?.text = self.date.customDateFormat
        }
    }
}

extension TestViewController: UIPickerViewDataSource {
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        
        return 10
    }
}

extension TestViewController: UIPickerViewDelegate {
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        
        return String(row)
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        
        self.value = row
        
        if let cell = self.tableView.cellForRow(at: 0) {
            
            cell.detailTextLabel?.text = String(self.value)
        }
    }
}


